crates/yttrium/src/pay/mod.rs

```diff
@@ -22,72 +33,831 @@ impl<T: std::fmt::Debug> From<progenitor_client::Error<T>> for PayError {
+    ) -> Result<GetPaymentResponse, PayError> {
+        let response = with_retry(|| async {
+            with_sdk_config!(self.client.gateway_get_payment(), &self.config)
+                .id(&payment_id)
+                .send()
+                .await
+        })
+        .await?;
+        Ok(response.into_inner())
+    }
+
+    /// Get payment options for given accounts
+    /// Also caches the options for use by get_required_payment_actions
+    pub async fn get_payment_options(
+        &self,
+        payment_link: String,
+        accounts: Vec<String>,
+    ) -> Result<PaymentOptionsResponse, GetPaymentOptionsError> {
+        let payment_id = extract_payment_id(&payment_link);
+        let body = types::GetPaymentOptionsRequest { accounts, refresh: None };
+        let response = with_retry(|| async {
+            with_sdk_config!(
+                self.client.gateway_get_payment_options(),
+                &self.config
+            )
+            .id(&payment_id)
+            .body(body.clone())
+            .send()
+            .await
+        })
+        .await
+        .map_err(map_payment_options_error)?;
+
+        let api_response = response.into_inner();
+
+        // Cache the options with their required actions
+        let cached: Vec<CachedPaymentOption> = api_response
+            .options
+            .iter()
+            .map(|o| CachedPaymentOption {
+                option_id: o.id.clone(),
+                required_actions: o
+                    .required_actions
+                    .iter()
+                    .cloned()
+                    .map(Into::into)
+                    .collect(),
+            })
+            .collect();
+        let mut cache = self.cached_options.write().map_err(|e| {
+            GetPaymentOptionsError::InternalError(format!(
+                "Cache lock poisoned: {}",
+                e
+            ))
+        })?;
+        *cache = cached;
+
+        Ok(api_response.into())
+    }
+
+    /// Get required payment actions for a selected option
+    /// Returns the list of actions from the cached options (must call get_payment_options first)
+    pub fn get_required_payment_actions(
+        &self,
+        option_id: String,
+    ) -> Result<Vec<RequiredAction>, GetPaymentRequestError> {
+        let cache = self.cached_options.read().map_err(|e| {
+            GetPaymentRequestError::InternalError(format!(
+                "Cache lock poisoned: {}",
+                e
+            ))
+        })?;
+
+        let cached_option = cache
+            .iter()
+            .find(|o| o.option_id == option_id)
+            .ok_or_else(|| {
+                GetPaymentRequestError::OptionNotFound(format!(
+                    "Option {} not found in cache. Call get_payment_options first.",
+                    option_id
+                ))
+            })?;
+
+        Ok(cached_option.required_actions.clone())
+    }
+
+    /// Fetch an action for a payment option
+    pub async fn fetch(
```

Comment:
Why do we need it as a public method?

We agreed that this should be called inside `get_required_payment_actions` only when payment action cannot be found for given payment option id from `cached_options`

Reply:
It's not exposed, LGTM